pkg/tlsx/openssl/openssl.go

```diff
@@ -0,0 +1,239 @@
+		Cipher:              tlsCipher,
+		TLSConnection:       "openssl",
+		CertificateResponse: c.convertCertificateToResponse(hostname, x509LeafCertificate),
+		ServerName:          serverName,
+	}
+	if c.options.TLSChain {
+		for _, opensslCert := range certificateChain {
+			x509Cert, err := c.convertOpenSSLToX509Certificate(opensslCert)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not convert openssl chain certificate")
+			}
+			response.Chain = append(response.Chain, c.convertCertificateToResponse(hostname, x509Cert))
+		}
+	}
+	return response, nil
+}
+
+func (c *Client) convertOpenSSLToX509Certificate(opensslCert *openssl.Certificate) (*x509.Certificate, error) {
+	pemBytes, err := opensslCert.MarshalPEM()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not marshal openssl to pem x509")
+	}
+	pemBlock, _ := pem.Decode(pemBytes)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not read openssl pem x509 to go pem")
+	}
+	if pemBlock.Type != "CERTIFICATE" {
+		return nil, errors.Wrap(err, "unsupported pem block type")
+	}
+	x509Certificate, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not convert openssl x509 to go x509")
+	}
+
+	return x509Certificate, nil
+}
+
+func (c *Client) convertCertificateToResponse(hostname string, cert *x509.Certificate) *clients.CertificateResponse {
+	response := &clients.CertificateResponse{
+		SubjectAN:  cert.DNSNames,
+		Emails:     cert.EmailAddresses,
+		NotBefore:  cert.NotBefore,
+		NotAfter:   cert.NotAfter,
+		Expired:    clients.IsExpired(cert.NotAfter),
+		SelfSigned: clients.IsSelfSigned(cert.AuthorityKeyId, cert.SubjectKeyId),
+		MisMatched: clients.IsMisMatchedCert(hostname, append(cert.DNSNames, cert.Subject.CommonName)),
+		IssuerCN:   cert.Issuer.CommonName,
+		IssuerOrg:  cert.Issuer.Organization,
+		SubjectCN:  cert.Subject.CommonName,
+		SubjectOrg: cert.Subject.Organization,
+		FingerprintHash: clients.CertificateResponseFingerprintHash{
+			MD5:    clients.MD5Fingerprint(cert.Raw),
+			SHA1:   clients.SHA1Fingerprint(cert.Raw),
+			SHA256: clients.SHA256Fingerprint(cert.Raw),
+		},
+	}
+	if parsedIssuer := parseASN1DNSequenceWithZpkix(cert.RawIssuer); parsedIssuer != "" {
+		response.IssuerDN = parsedIssuer
+	} else {
+		response.IssuerDN = cert.Issuer.String()
+	}
+	if parsedSubject := parseASN1DNSequenceWithZpkix(cert.RawSubject); parsedSubject != "" {
+		response.SubjectDN = parsedSubject
+	} else {
+		response.SubjectDN = cert.Subject.String()
+	}
+	if c.options.Cert {
+		response.Certificate = clients.PemEncode(cert.Raw)
+	}
+	return response
+}
+
+// parseASN1DNSequenceWithZpkix tries to parse raw ASN1 of a TLS DN with zpkix and
+// zasn1 library which includes additional information not parsed by go standard
+// library which may be useful.
+//
+// If the parsing fails, a blank string is returned and the standard library data is used.
+func parseASN1DNSequenceWithZpkix(data []byte) string {
```

Comment:
we can make this a common function within clients package and reuse it for both stdlib and openssl implementation along with the above logic.

```
if parsedIssuer := parseASN1DNSequenceWithZpkix(cert.RawIssuer); parsedIssuer != "" {
		response.IssuerDN = parsedIssuer
	} else {
		response.IssuerDN = cert.Issuer.String()
	}
	if parsedSubject := parseASN1DNSequenceWithZpkix(cert.RawSubject); parsedSubject != "" {
		response.SubjectDN = parsedSubject
	} else {
		response.SubjectDN = cert.Subject.String()
	}
```

Reply:
done